scsynth: improve audio thread priority setting on macOS

dyfer opened the pull request:
<!-- For information about contributing see: https://github.com/supercollider/supercollider/wiki/Contributing-directory -->

## Purpose and Motivation

<!-- Please provide a description, even if you are linking to a related Issue or PR. -->
<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->

I observed, that it is somewhat easy to create dropouts in scsynth when e.g. application windows are moved. That makes me think that maybe the thread priorities are not set appropriately.

I tried to leverage the `thread_set_priority_rt()` function for this. It has _slightly_ improved the situation on my system (M1pro). Test code below.

I'm not sure if this is the best sollution, or if the `computation` and `constraint` values should be adjusted further.

<details><summary>Example code I used for stressing scsynth</summary>

```supercollider
(
Server.scsynth;
s.options.memSize_(2 ** 20); // needed for the load test
s.options.hardwareBufferSize_(128);

~numSynths = 500;
s.waitForBoot{
	SynthDef(\sinVerb, { arg out = 0, amp = 0.01;
		Out.ar(out, GVerb.ar(SinOsc.ar(Rand(400, 700))) * amp);
	}).add;

	~allSynths ?? {~allSynths = []};
	~allSynths = ~allSynths.addAll(
		~numSynths.collect({
			s.sync;
			Synth(\sinVerb, [\amp, -60.dbamp]);
		})

	)
}
)

~allSynths.do(_.free)

// macOS 14, CPU Apple M1pro, 500 synths
// this PR: peak around 90-100%, avg around 72%; no or minimal dropouts when moving windows
// 3.13: peak around 90-100%, avg around 85%; no dropouts when idle; dropouts when moving windows
```

</details>

Regarding the implementation: note that I had to move the thread creation to after the driver was initialized - this was needed in order to get the samplerate, which is needed to set the deadlines.

EDIT:
I have also tested this on a 2015 MacBook Air (Intel). 
For 100 synths from the example, SC 3.13 produced constant dropouts; this build produced only minimal and occasional dropouts.

EDIT 2:
~~Note that before SC 3.14 we were inadvertently dropping the thread priority for the audio thread...~~ We were dropping the thread NRT thread, apparently. See https://github.com/supercollider/supercollider/pull/6043. 

## Types of changes

<!-- Delete lines that don't apply -->

- New feature (?)

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [ ] All tests are passing
- [ ] Updated documentation
- [x] This PR is ready for review <!-- If not ready for review, consider making a Draft PR instead, or describe what kind of specific feedback you want. -->

<!-- ## Merge notes
(optional) You may also add notes of:
     - Dependencies: if this PR depends on any other PRs or actions
     - Merge Note: any considerations regarding merging this PR
     - Release Notes: text you think would be useful to include in the Release Notes
-->

<!-- Note: The recommended commit message format is:
     component: sub-component: short description of changes
For example:
     classlib: ClassBrowser: fix search with empty query string
Common `comp:` labels:
     help, classlib, sclang, scsynth, plugins, tests, docs, SCDoc
See wiki for more info: Wiki->Creating Pull Requests->git-commit-message-format
-->

telephon commented:
> Btw this PR also made a difference when testing on older intel machine. On the surface it seems like a desirable feature, but I'm not very familiar with the intricacies of scheduling etc. It would be great to get feedback from someone who has more experience with it, especially on macOS.

I haven't seen the issue on my intel mac so far, but I'll be running this branch for a while to see if I notice anything positive (or negative).

dyfer commented:
Thanks @telephon 
Just to be clear: the behavior that I observed is when reaching the max CPU load and slight dropouts start to appear. With this PR I observed the slight dropouts to be minimized, both on M1 and Intel chips.

dyfer commented:
Also see #7009 for a possible sclang timing improvement.

redFrik commented:
i did some tests and came to the conclusion that the SuperCollider 3.13.0 i had (Built from tag 'Version-3.13.0' [3188503]) was the problem.  it is nearly unusable on my M1 Pro Sonoma - lots of timing glitches - even under the lightest of loads.
test code was basically just a couple of Pbinds with server latency 0.05, sampleRate 48K, blockSize 64.

with SuperCollider 3.14.0-rc1 i cannot make the same code produce any late messages. also this PR and [#7009](https://github.com/supercollider/supercollider/pull/7009) feel stable.
with your extreme examples i notices differences, but they're subtle.

so my big issue all along was that 3.13.0 build.  thanks and sorry for diverting the discussion.

dyfer commented:
> with SuperCollider 3.14.0-rc1 i cannot make the same code produce any late messages

Yes, this was changed in https://github.com/supercollider/supercollider/pull/6043. However, that removed raising the priorities... which this PR tries addresses.

BTW I noticed that PR also removed raising the priority for the resync thread, I added it here.

Spacechild1 commented on server/scsynth/SC_CoreAudio.cpp:
```diff
@@ -339,8 +347,26 @@ SC_AudioDriver::~SC_AudioDriver() {
+    // printf("Verifying values: mSampleRate = %f, blockSize = %d, period = %d, computation = %d, constraint = %d\n",
+    // mSampleRate, blockSize, ns_per_block, computation, constraint);
+
+    bool success = nova::thread_set_priority_rt(ns_per_block, computation, constraint, true);
```
As this is the NRT thread, we want a thread priority that is high, but still below the audio thread. This is what `int priority = nova::thread_priority_interval_rt().first;` attempts to do on Linux and Windows. We need to do a similar thing on macOS. AFAICT, your code essentially sets the same priority as the audio thread, which doesn't seem right.

dyfer replied:
Thanks! Where's the audio thread? What's its priority?

This implementation definitely improves scsynth on macOS (less dropouts / more headroom) - I'm not saying it's right but I'm reinforcing that it did _something_.

Spacechild1 commented:
@timblechmann How would you set thread priorities on macOS in 2025? We need a mid-high priority for the NRT and network thread and high priority for Supernova's DSP helper threads.

dyfer commented:
> No, this was about the _NRT thread_.

Sorry for my misunderstanding!

dyfer commented:
Hm... on testing today, I can't see the difference between the current develop build and this PR. Maybe I was chasing a red herring...

Spacechild1 commented:
> Maybe I was chasing a red herring...

Yeah, I can't see how this PR could possible reduce audio dropouts, so the likely explanation is that it doesn't :)

dyfer commented:
Sorry for the noise, everybody. 
I was mistakenly comparing this against SC 3.13 (which _is_ more prone to dropouts on M1).

Closing for now.
